### ISSUE TITLE
Semantics of @.foo as a standalone expression (truthiness)

### ISSUE THREAD

**Issue by gregsdennis** [ietf-wg-jsonpath/draft-ietf-jsonpath-base#119]
Currently we have `$[?(@.foo)]` (I'm just going to use `@.foo` for brevity) to represent "the item is an object and contains a `foo` property."  It also seems that some implementations evaluate the truthiness of the value of the `foo` property, if it exists, while others that accept the syntax interpret it to strictly consider existence.

Related comparison results:
- [Filter Expression On Object](https://cburgmer.github.io/json-path-comparison/results/filter_expression_on_object.html)
- [Filter Expression Without a Value](https://cburgmer.github.io/json-path-comparison/results/filter_expression_without_value.html)

There has also been an [email thread](https://mailarchive.ietf.org/arch/msg/jsonpath/yJEZ1bbdFt19qRAeHEoMqtB6jRg/) floating around discussing truthiness, but I wanted something a bit easier to read through than emails.

One aspect that I don't think has been considered in the email thread is the ability to consider strict existence.  On StackOverflow, [this](https://stackoverflow.com/questions/69069083/write-a-jsonpath-query-for-a-null-property-with-newtonsoft-json-net) and [this](https://stackoverflow.com/questions/69120789/how-to-get-the-result-which-has-not-the-property-with-jpath-in-the-newtonsoft) question both have asked about Newtonsoft's (.Net) support for this syntax, and the second seems to be asking specifically about detecting the _absense_ of a property as opposed to its presence with a `null` value.

I think strict existence detection is an important feature to support, and if we move to have this syntax operate on truthiness, I fear we won't be able to support it.

Importantly, I would suggest that regardless of the decision we make here, `!@.foo` should return the inverse set.

**Comment by glyn** [comment]
> `!@.foo` should return the inverse set.

Yes! `@.foo` should always be `true` or `false` and `!@.foo` should always be the negation of `@.foo`.

I would prefer:
* `@.foo` always to denote strict existence of the key `foo` in the object `@` (and be `false` in other cases, e.g. when `@` is not an object).
* there to be no implicit conversion of values to booleans and therefore no need for the concepts of truthy and falsy in the spec.

**Comment by cabo** [comment]
> Yes! `@.foo` should always be `true` or `false` 

You lost me here.

`@.foo.bar` doesn't work any longer if `@.foo` is defined that way.

> and `!@.foo` should always be the negation of `@.foo`.

That is one of the places where @.foo could be used in a context that expects a Boolean value, as is `$[?(@.foo)]`.
I think you are saying that there should be an implicit conversion of the range of @.foo (which I called a nodelist before, but we don't need to stick with that) to Boolean.

And then you are outlining what that conversion should look like:

> * `@.foo` always to denote strict existence of the key `foo` in the object `@` (and be `false` in other cases, e.g. when `@` is not an object).

Which means that I don't understand this:

> * there to be no implicit conversion of values to booleans and therefore no need for the concepts of truthy and falsy in the spec.

**Comment by remorhaz** [comment]
`@.foo` as itself just provides some node list (either empty or not). Then we can:

- Use it in filter directly or with logical operator, like `$[?(@.foo)]`, or `$[?(@.foo and @.bar == 1)]`, or `$[?(!@.foo)]`. In this cases it's check for existence (non-emptiness of node list).
- Use it in filter as a part of comparison, like `$[?(@.foo == true)]`. Comparison evaluates to `true` only if both of his part exist and provide equal value.

> ...one of the places where `@.foo` could be used in a context that expects a Boolean value, as is `$[?(@.foo)]`.

There are indeed contexts that require casting to boolean, like "filter root" or "logic operator argument" (but not "comparison operator argument"). And node list casts to `false` if it's empty and to `true` otherwise - I see it like this.

**Comment by cabo** [comment]
> * Use it in filter as a part of comparison, like `$[?(@.foo == true)]`. Comparison evaluates to `true` only if both of his part exist and provide equal value.

What is the semantics of `$[?(@.foo != true)]`?
Is that the same as `$[?(!(@.foo == true))]`?

**Comment by glyn** [comment]
Sorry @cabo: I was using `@.foo` in the same way as the OP:

> Currently we have `$[?(@.foo)]` (I'm just going to use `@.foo` for brevity) to represent "the item is an object and contains a foo property." It also seems that some implementations evaluate the truthiness of the value of the foo property, if it exists, while others that accept the syntax interpret it to strictly consider existence.

I prefer the latter interpretation. To be clearer, I would be happy to see the syntax in the current draft amended to something like:
```
exist-expr   = *neg_op (root-selector / "@") 1*(dot-selector / ("[" quoted-member-name "]"))
```
Thus attempts to use other expressions (such as `@` and `$[0]`) as existence tests would be syntactically invalid.

**Comment by remorhaz** [comment]
> Thus attempts to use other expressions (such as @ and $[0]) as existence tests would be syntactically invalid.

What rationale is behind forbidding to test existance of `$`-based paths? We can have complex filters like this: `$.a[?(@.foo == true and $.b)]` that selects `$.a` elements with `"foo": true` or nothing if `$.b` doesn't exist.

**Comment by gregsdennis** [comment]
This issue has expanded into more than what was intended.  So let me clarify my intents:

- By `@.foo` in expressions, I mean to include any path as used in the context of an expression, be it `@.foo`, `@.foo.bar`, `@[0]`, `$.foo`, or any other path.  I'm not asking whether these are valid in the context of an expression.  I would expect from previous conversations and what's in the spec currently that we agree that they are.
- We've also previously discussed that such paths within expressions must return a single value.  This is not up for discussion here.
- I only intend to discuss the usage of a path in an expression when it is not part of an equality or other comparison.  Specifically, when the only operators, if any, that operate on the path are boolean operators (i.e. `!`, `&&`, and `||`).
- The result set is defined in the current incarnation of the spec to be a nodelist.  This was added out of a need to define it separately from any other JSON array, and it really only applies when discussing the theory behind the spec.  Implementations can design their internal mechanisms however they want so long as the output is what is prescribed by the specification.  This is not up for discussion.

With that in mind, let's continue the discussion at hand, please.

I'm asking about considering the results of such expressions as "truthy."  It was my opinion (upon opening the issue) that by considering the result set to be truthy, we restrict our ability to check for strict existence.

Mulling this over for a bit now, I see three options.

1. Consider paths in expressions as truthy.  This means that we need to define a way to query for strict existence.  We could introduce a function such as `exists(@.foo)` to handle this, but function syntax has been deferred a few times because we haven't defined a sufficient foundation to discuss them.  This seems to also be in line with many existing implementations, though no clear consensus exists.
2. Consider paths in expressions as a strict check for existence.  This means that truthy queries be more complex to include the full intent of truthiness.  This raises questions around what that looks like.  Do we consider `@.foo == true` to be sufficient to express truthiness?  We already decided that `==` and kin don't perform type conversions, so that seems to be off the table.  But that means a user would have to list out all of the "truthiness" possibilities.  This gets very messy very quickly, which leads us to the last option.
3. Add a separate operator set for truthy comparisons vs strict comparisons.  This moves us more toward a Javascript operator set instead of the C operator set we've decided on so far.  The nuances of this are less widely known, which means it becomes harder to adopt JSON Path.

Considering the implications of these options has led me to believe that the first one is actually most viable.  We just need to define how functions work.

---

As an alternative to all of this, I would be happy to have the specification require the implementation to provide an option to consider such expressions as either truthy or strict existence, and then define what each means.  This would mean a little more work for implementers, but I am one, and I can tell you that I don't mind.  It's our end users that we need to consider most.

**Comment by remorhaz** [comment]
What I was speaking about before was exactly option 2.

**Comment by cabo** [comment]
> * I only intend to discuss the usage of a path in an expression when it is not part of an equality or other comparison.  Specifically, when the only operators, if any, that operate on the path are boolean operators (i.e. `!`, `&&`, and `||`).

This was what I meant with "in a context where a Boolean is expected".  So we are still talking about implicit conversions of query results to Boolean.

> 2. Consider paths in expressions as a strict check for existence.  

This.

> This means that truthy queries be more complex to include the full intent of truthiness.  

You are presupposing that we need another implicit conversion to Boolean, called "truthy".
I haven't seen evidence that we need that.

> This raises questions around what that looks like.  Do we consider `@.foo == true` to be sufficient to express truthiness?  

Of course not (== true is the exact opposite of "truthy").  There is an implicit conversion of a query result to a JSON value involved here that we should discuss separately.

> We already decided that `==` and kin don't perform type conversions, so that seems to be off the table.  But that means a user would have to list out all of the "truthiness" possibilities.  This gets very messy very quickly, which leads us to the last option.

Again, where is the evidence that "truthy" makes sense?  And which of the many ways of defining it?

> 3. Add a separate operator set for truthy comparisons vs strict comparisons.  This moves us more toward a Javascript operator set instead of the C operator set we've decided on so far.  The nuances of this are less widely known, which means it becomes harder to adopt JSON Path.

More likely, an explicit coercion from a JSON value to a Boolean along the "truthy" lines would be indicated.

`truthy(@.foo)` if we must.
But leaving this out is better.

**Comment by goessner** [comment]
I do like Glyn's syntax
```
exist-expr   = *neg_op (root-selector / "@") 1*(dot-selector / ("[" quoted-member-name "]"))
```
In fact it already includes the root-selector `$`. But considering an argument
```
[ { a:2 } ]
```
we obviously need to support queries with array element existence checks as in

`$[?(@.a && $[1])]`

And ... don't we already have, beside those implicit existence tests above, with

`'foo' in @`

an explicit existence test instead of inventing an `exist()` function?

**Comment by danielaparker** [comment]
> `@.foo` should always be `true` or `false` and `!@.foo` should always be the negation of `@.foo`.
> 
> I would prefer:
> 
> * `@.foo` always to denote strict existence of the key `foo` in the object `@` (and be `false` in other cases, e.g. when `@` is not an object).
> * there to be no implicit conversion of values to booleans and therefore no need for the concepts of truthy and falsy in the spec.

I believe that would be well defined. The only issue with it is that it would be inconsistent with almost all (I added "almost" because I haven't looked at all of them), so "almost all", existing implementations.

Daniel

**Comment by cabo** [comment]
@danielaparker:  So no `@.foo.bar` and no `@.foo == 17`?

**Comment by remorhaz** [comment]
> So no `@.foo.bar` and no `@.foo == 17`?

But why? Because of "`@.foo` is always boolean" in @glyn 's phrase?

**Comment by danielaparker** [comment]
> @danielaparker: So no `@.foo.bar` and no `@.foo == 17`?

Consider the JSON document
```
[{"foo" : "baz"}]
```
Glyn's rule (assuming I've understood it correctly) would imply that
```
$[?(@.bar == false)]
```
would produce
```
[{"foo" : "baz"}]
```
I don't know of any implementation that does so.

**Comment by glyn** [comment]
Please note my syntax fragment only applied to `exist-expr` and didn't limit the terms of comparisons etc.

**Comment by danielaparker** [comment]
> Please note my syntax fragment only applied to `exist-expr` and didn't limit the terms of comparisons etc.

But if the evaluation of `@.bar` by itself should differ from when it appears in a binary expression, that would also be different from "almost all" existing implementations. 

I think "almost all" existing implementations take one of three approaches to evaluating an object with a missing key:

- Return a valid JSON value, generally null. Truthiness rules will produce false when testing a single value. The equality operator will evaluate to true when comparing with the JSON value `null`.  
- Return a special non JSON value, in JavaScript this is `undefined` Truthiness rules will produce false when testing a single value. The equality operator will evaluate to false when comparing with any JSON value. The equality operator will evaluate to true when comparing with the special value itself, e.g. `undefined` in JavaScript. 
- Treat a missing property in an object as a runtime type error. Then the question is how the implementation handles the type error. It can get complicated.    

Jayway JsonPath is an example of the third approach, and it handles this type error differently depending on where in the path it occurs. Evaluating `$.bar`, with missing 'bar', it will throw a `PathNotFound` exception with the message ""Missing property in path". Evaluating `$[?(@.bar || @.foo )]` with missing `bar` and present `foo`, branches to some code that ignores that part of the path evaluation so that other parts of the path evaluation can still be examined.

**Comment by glyn** [comment]
> But if the evaluation of @.bar by itself should differ from when it appears in a binary expression, that would also be different from "almost all" existing implementations.

Thanks for the reality check. I don't like truthy and falsey, but it looks like we are stuck with them.

**Comment by goessner** [comment]
Please remember, that JSONPath is about selecting nodes. In the following cases of non-existence of object member `foo`  we simply do not need to return its value:

1. `$.foo` ... the resulting node-list is empty.
2. `$.foo.bar` ... the resulting node-list is empty.
3. `$[?(@foo)]` ...  the resulting node-list is empty.
4.  `$[?(!@foo)]` ... current `@` is in the node-list.

Only temporary boolean values are needed with boolean expressions

5.  `$[?(!!@foo)]` ...  the resulting node-list is empty.
6.  `$[?(!@foo || @bar && $.baz)]` ...  the resulting node-list is empty or contains `@`

But now with direct comparisons we need to convert its non-existence to a (boolean ?) value.

7.  `$[?(@foo==false)]` ... the resulting node-list is empty ... regardless of the right side
8.  `$[?(@foo!=false)]` ... hmm ... to be defined.
9. `$[?!(@foo==false)]` ... current `@` is in the node-list.

BTW: This is, where I like the mental node-list approach. I don't need to specify the selection process result as values or normalized pathes, despite the fact, that it will be one of them.

A JSON value `undefined` would be of help here ... but since that value in above cases is not exposed to the outer world, it is up to the implementation to represent it internally somehow specifically. This seems to conform to @danielaparker's case 2 above.

The spec requires to only treat syntax errors. No 'type not found' message is required or even recommended. It is up to the implementations to handle such cases.

**Comment by danielaparker** [comment]
> In the following cases of non-existence of object member `foo` we simply do not need to return its value:

> But now with direct comparisons we need to convert its non-existence to a (boolean ?) value.
> 
I don't think that's a useful way of presenting the problem.

Generally, as an expression evaluator evaluates the parse tree, it needs to evaluate the operands _before_ performing any unary or binary operations, including comparisons. Therefore, it needs to produce _something_ to represent the result of evaluating `@.foo`.  If we're using the Shunting-yard algorithm, for example, that something has to be pushed onto an output queue of values. In my case 1, that something would be a JSON null value; in my case 2, it would be an `undefined` value.  Another possibility would be an 'optional' type value with properties `HasValue` and `Value`. Of course, there are implementation details, but the basic point remains the same: how we think about `@.foo` doesn't change as we make our way through the parse tree.

**Comment by danielaparker** [comment]
> The spec requires to only treat syntax errors. No 'type not found' message is required or even recommended. It is up to the implementations to handle such cases.

It still behooves the specification, for each expression operator, to identify the runtime type errors that may occur, and how the response should be handled. For example, if the ordering operators `<`, '<=', '>', and '>=' were specified to be only valid for numbers, as in JMESPath, then other types would entail a type error. Possible responses could be to take out just the element currently being evaluated from the result set (as JMESPath specifies for this case), or to return an entirely empty collection to the user.

**Comment by gregsdennis** [comment]
I've updated my library to use strict existence.  You can play with it at [json-everything.net/json-path](https://json-everything.net/json-path).

Here's an example:

Path
```
$[?(!@.name || @.name == 'abc')]
```

Data
```json
[
  {
      "id": 1,
      "name": "abc"
  },
  {
      "id": 2
  },
  {
      "id": 3,
      "name": null
  },
  {
      "id": 4,
      "name": "def"
  }
]
```

Response
```
[
    {"id":1,"name":"abc"},
    {"id":2}
]
```

Note that the third item (id: 3) is not selected.  If we use truthiness, then this would be selected and there would be no way to selecting either the second or third item without selecting the other.

**Comment by cabo** [comment]
Great.  So with Greg's implementation, we now have two different implicit conversions:

- using the result of a query expression in a Boolean context (?(), !, &&/||): empty node sets ➔ false, non-empty ➔ true
- using the result of a query expression in a comparison (and I expect anywhere else): retrieve the value at "the node"

I think we don't have a full grasp of the second alternative.  What if the query expression would have been `$..foo` (and there is more than one node in the nodeset)?
What does it mean to compare an empty nodeset to anything?  Do we get `a == b` ≣ `!( a != b)`?

**Comment by cabo** [comment]
Meta-comment: Obviously, I'd prefer to get rid of the implicit conversion (e.g., always use something like `exist(@.name)` as an explicit existence test).
But we want to support current practice here. 
I think we need to decide what current practice we can cover and what goes beyond what we can do.
I think Greg's list of

- [Filter Expression On Object](https://cburgmer.github.io/json-path-comparison/results/filter_expression_on_object.html)
- [Filter Expression Without a Value](https://cburgmer.github.io/json-path-comparison/results/filter_expression_without_value.html)

and the "Proposal A" in there is a very good start.

**Comment by danielaparker** [comment]
> If we use truthiness, then this would be selected and there would be no way to select either the second or third item without selecting the other.

@gregsdennis , You clearly don't understand what "truthiness" means :-) Could I suggest reading this short section from Crafting Interpreters? https://craftinginterpreters.com/evaluating-expressions.html#truthiness-and-falsiness? Whether 
`{"id": 3, "name": null}` is selected depends on the rules the language designer chooses for determining whether an expression is `falsey` or `truthy`.  Different languages have adopted different rules. Different JsonPath implementations have adopted different rules. You can of course have a view about what those rules ought to be. But the statement "If we use truthiness, then this would be selected" is nonsensical.

Best regards,
Daniel

**Comment by gregsdennis** [comment]
I think you found a bug in my implementation.  I definitely don't intend to select `name : false`.  I'll fix it later tonight and redeploy.

**Comment by danielaparker** [comment]
> I think you found a bug in my implementation. I definitely don't intend to select `name : false`. I'll fix it later tonight and redeploy.

I think that result is consistent with how you've defined what `@.name` means. And it looks from the comparisons that at least six implementations, including the important Jayway, have the same behaviour. I've submitted a pull request to the comparisons to get a more accurate count.

**Comment by gregsdennis** [comment]
🤔 Well, that was not my intent.  I want `!@.name` to select an item _only_ if `name` doesn't exist.  This means a user would be forced to write `@.name == false` to match your case.

I guess this is an option we can throw into the mix:

1. `@.foo` is fully truthy, so `foo` missing or present with a value of `null`, `false`, `0`, `""`, `{}, or `[]` doesn't match; any other value does match.
2. `@.foo` is an existence test, so `foo` missing doesn't match; any value at all does match.  (what I intended)
3. `@.foo` is an existence test, but its value can matter _if it's a boolean_, so `foo` missing or present with a value of `false` doesn't match; any other value does match.  (what mine currently does)

**Comment by danielaparker** [comment]
>>>I was interested in the case of document
```
[{"name": false}]
```
with query
```
$[?(@.name)]
```
Your implementation returns

{"name":false}

> > I definitely don't intend to select `name : false`. 
> 
> I think that result is consistent with how you've defined what `@.name` means. I expected it. What I didn't expect was that at least six implementations, including the important Jayway, have the same behaviour. I've submitted a pull request to the JSONPath Comparisons to get a more accurate count.

And [here](https://cburgmer.github.io/json-path-comparison/results/filter_expression_without_value.html) we have the results. 

To summarize: 14 of 42 implementations will return `{"name":false}`. These include the Jayway implementation, which is probably second to Goessner as far as importance for JSONPath. Unsurprisingly, it excludes the implementations including Goessner JavaScript that use a dynamic language such as JavaScript for evaluating expressions.

**Comment by danielaparker** [comment]
> > But if the evaluation of @.bar by itself should differ from when it appears in a binary expression, that would also be different from "almost all" existing implementations.
> 
> Thanks for the reality check. I don't like truthy and falsey, but it looks like we are stuck with them.

@glyn, It appears from [these results](https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/119#issuecomment-918148025) that my remark that you're replying to is wrong. In fact 14 of 42 implementations do evaluate `@.bar` by itself differently than when it appears in a binary expression.

**Comment by danielaparker** [comment]
Below is a table that shows all of the falsey conditions for each of the 42  implementations in the JSONPath Comparisons, deduced from [this query](https://cburgmer.github.io/json-path-comparison/results/filter_expression_without_value.html). These are tests of the query `$[?(@.key)]` against a JSON array containing an assortment of objects containing a single name-value pair.

A checkmark appears in the column labelled "n/a" if an implementation fails to execute the query  `$[?(@.key)]` or returns an empty result set. A checkmark appears in the column labelled "Key exists" if an implementation selects all objects having member name "key". The remaining columns show various groupings of falsey conditions. The checked column applies to the implementation on that row.

Implementation|n/a|Key exists|false|null|false, null|false, null, "", [], {}|false, null, "", [], {}, 0|false, null, string, [], {} , 0|false, null, 0, ""|false, null, 0, -ve|false, null, 0, ""|false, null, 0, "", []|
--------------|-----------|-------|--------------|-------------|--------------------|----------------------------|------------------------------|----------------------------------|------------------------|----------------------------|---------------------------|-------------------------------|
Bash (JSONPath.sh)                    |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
C (json-glib)                         |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Dart (json_path)                      |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Elixir (ExJsonPath)                   |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Golang (github.com-ohler55-ojg)       |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Golang (github.com-oliveagle-jsonpath)|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Haskell (jsonpath)                    |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Kotlin (com.nfeld.jsonpathkt)         |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Python (jsonpath-ng)                  |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Python (jsonpath-rw)                  |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Raku (JSON-Path)                      |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Rust (jsonpath)                       |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
PHP (Goessner)                        |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Perl (JSON-Path)                      |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
dotNET (JsonPathLib)                  |<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Elixir (warpath)                              |&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Erlang (ejsonpath)                            |&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Golang (github.com-vmware-labs-yaml-jsonpath) |&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Java (com.github.jsurfer)                     |&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Java (com.jayway.jsonpath)                    |&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Objective-C (SMJJSONPath)                     |&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
PHP (remorhaz-jsonpath)                       |&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Rust (jsonpath_lib)                           |&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Scala (jsonpath)                              |&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
dotNET (Json.NET)                             |&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
dotNET (JsonPath.Net)                         |&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
PHP (galbar-jsonpath)                         |&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;| 
dotNET (Manatee.Json)                         |&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Clojure (json-path)                   |&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Ruby (jsonpath)                               |&nbsp;|&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;| 
Cpp (jsoncons)                                |&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;| 
dotNET (JsonCons.JsonPath)                    |&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;| 
Golang (github.com-spyzhov-ajson)    |&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;| 
Python (jsonpath)                    |&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;| 
Golang (github.com-PaesslerAG-jsonpath) |&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|&nbsp;|
Golang (github.com-bhmj-jsonslice)             |&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|
Golang (github.com-bhmj-jsonslice)             |&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|&nbsp;|&nbsp;|
JavaScript (Goessner)                   |&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|    
JavaScript (brunerd)                    |&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|
JavaScript (jsonpath)                   |&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|
JavaScript (jsonpath-plus)              |&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|&nbsp;|
PHP (softcreatr-jsonpath)               |&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|&nbsp;|<em>&#x2713;</em>|

**Comment by cabo** [comment]
Fixed in 76789d85

**Comment by gregsdennis** [comment]
Can you summarize the resolution, please?

**Comment by cabo** [comment]
Since 76789d8, Section 3.5.9 now says:

```
   *  A member or element value by itself in a Boolean context is
      interpreted as false only if it does not exist.  Otherwise it is
      interpreted as true.  To be more specific about the actual value,
      explicit comparisons are necessary.  This existence test -- as an
      exception to the general rule -- also works with structured
      values.
```
